### PR TITLE
Fire - Noor

### PR DESCRIPTION
**Pull request by hn4ever**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The model is where all the databases of the rail app are located.
Describe in your own words what the Controller is doing in Rails | The controller receives events like HTTP requests, gather data from model, and pass the data to be viewed as HTML.
Describe in your own words what the View is doing in Rails | The view is the resulted HTML page that has display logic.
Describe an edge-case controller test you wrote | The "Toggle complete" test was an edge case, I tried to mark a task as complete, then marked as incomplete and later I made a test  to check if the value of "completion_at" is nil
What is the purpose of using strong params? (i.e. the params method in the controller) | It's used inside a controller in order to safely acquire data from forms.
How are Rails migrations related to Rails models? | Model is where the database is and with migration we can have more control of the database in rails and it will also be useful for testing.
Describe one area of Rails that are still unclear on | I'm still confused with how the routes should be organized and about HTTP verbs.

### CODE REVIEW

File: test/controllers/tasks_controller_test.rb [hunk 1]
```diff
@@ -88,37 +84,89 @@
...
+    end
+    let (:new_task_hash) {
+      {
+        task: {
+          name: "Buy a Book",
+          description: "Buy a book about six sigma",
+          completed_at: "today"
+        },
+      }
+    }
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:116]: Great work leveraging the `let` syntax!

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -88,37 +84,89 @@
...
+      task = Task.find_by(id: task.id)
+      expect(task.name).must_equal new_task_hash[:task][:name]
+      expect(task.description).must_equal new_task_hash[:task][:description]
+      expect(task.completed_at).must_equal new_task_hash[:task][:completed_at]
     end
     
     it "will redirect to the root page if given an invalid id" do
-      # Your code here
+      get update_task_path(-1)
+      must_redirect_to root_path
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:134]: Well-written and concise test :)

File: test/controllers/tasks_controller_test.rb [hunk 3]
```diff
@@ -88,37 +84,89 @@
...
+    
+    patch uncomplete_task_path(task.id)
+    task = Task.find_by(id: task.id)
+    expect(task.completed_at).must_be_nil
+
+    patch complete_task_path(task.id)
+    task = Task.find_by(id: task.id)
+    expect(task.completed_at).wont_be_nil
+    
+    end
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:170]: Watch your indentation!

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    def new
+        @task = Task.new
+    end
+
+    def create
+        @task = Task.new(
+            name: params[:task][:name],
+            description: params[:task][:description],
+            completed_at: params[:task][:completed_at]
+        )
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:25]: It wasn't a requirement for this project, but if you had more time on this project, I'd encourage you to use a strong params method!

File: app/controllers/tasks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    def uncomplete
+        @task = Task.find(params[:id])
+        update = @task.update(completed_at: nil)
+        redirect_to root_path
+    end
+
+    def count 
+        count = Task.count
+    end
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:80]: Does this method get used? If I comment this method out, all of your tests still pass.

If this method doesn't get used, then it's better to delete it!

File: test/controllers/tasks_controller_test.rb [hunk 6]
```diff
@@ -88,37 +84,89 @@
...
+      completed_at: "today"
+    )
+    
+    patch uncomplete_task_path(task.id)
+    task = Task.find_by(id: task.id)
+    expect(task.completed_at).must_be_nil
+
+    patch complete_task_path(task.id)
+    task = Task.find_by(id: task.id)
+    expect(task.completed_at).wont_be_nil
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:168]: I love this test! I think it's pretty awesome that you did multiple "act"s and multiple "assert"s. Usually I'd like to discourage that, but in this case, the asserts are very related, so it reads very well. :)

**Review by tildeee — COMMENTED** [review]
# Task List




## Major Learning Goals/Code Review





| Criteria                                           | yes/no, and optionally any details/lines of code to reference |
| -------------------------------------------------- | ------------------------------------------------------------- |
| At least 6 commits with meaningful commit messages | ✔️                                                            |
| Routes follow RESTful conventions                  | ✔️                                                            |
| Uses named routes (like `_path`)                   | ✔️                                                            |
| Creates Models and migrations                      | ✔️                                                            |
| Creates styled views                               | ✔️                                                            |
| Handles errors like nonexistant tasks              | ✔️                                                            |
| Uses `form_with` to render forms in Rails          | You ended up using `form_for` instead, which is valid and works, but for this project we were expecting `form_with`                                                            |

## Functional Requirements/Manual Testing

| Functional Requirement                                                         | yes/no |
| ------------------------------------------------------------------------------ | ------ |
| Successfully handles index & show                                              | ✔️     |
| index & show tests pass                                                        | ✔️     |
| Successfully handles: New, Create                                              | ✔️     |
| New, Create tests pass                                                         | ✔️     |
| Successfully handles: Edit, Update                                             | ✔️     |
| Edit, Update tests pass with valid & invalid task ids                          | ✔️     |
| Successfully handles: Destroy, Task Complete                                   | ✔️     |
| Tests for Destroy & Task Complete include tests for valid and invalid task ids | No test for Task Complete with invalid task id     |

## Overall Feedback

| Overall Feedback                | Criteria                                                                                                                                                                                      | yes/no |
| ------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 6+ in Functional Requirements                                                                                                                                            | ✔️
| Yellow (Approaches Standards)   | 3+ in Code Review && 5+ in Functional Requirements, or the instructor judges that this project needs special attention                                                                        |
| Red (Not at Standard)           | 0-2 in Code Review or 0-4 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |

### Additional Feedback

Great work on this project, Noor! Your project has all the functionality I was looking for.

Your tests also look very good! I like how they're written. And I like the styling you put on your site :) !!!

There are a couple of things I'm commenting on your code-- nothing to do except think about them and let me know if you have any questions.

Also, something you should note: When creating a task, it already says that it is complete. I wouldn't expect that functionality, so it'd be interesting to figure out why it's happening!

Overall, great work!

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality              | Yes? |
| -------------------- | ---- |
| Perfect Indentation  |     |
| Elegant/Clever       |     |
| Descriptive/Readable |     |
| Concise              | ✅    |
| Logical/Organized    |     |

File: test/controllers/tasks_controller_test.rb (hunk 1 above)
Comment (hn4ever) [reply test/controllers/tasks_controller_test.rb:116]: Thank you Simon

**Comment by hn4ever** [comment]
Thank you Simon
